### ISSUE TITLE
packer validate failures when using the data "sshkey" source 

### ISSUE THREAD

**Issue by puetzk** [hashicorp/packer#11294]
When filing a bug, please include the following headings if possible. Any
example text in this template can be deleted.

#### Overview of the Issue

I originally reported this as https://github.com/ivoronin/packer-plugin-sshkey/issues/3, but it seems to be an issue in packer itself; there doesn't seem to be anything a datasource plugin can do to make validate work.

#### Reproduction Steps

packer validate sshkey.pkr.hcl` fails with

> Error: 1 error(s) occurred:
> 
> * ssh_private_key_file is invalid: CreateFile <unknown>: The filename, directory name, or volume label syntax is incorrect.
> 
>   on sshkey.pkr.hcl line 12:
>   (source code not available)

(the same is true for real builders, but this minimal example shows the issue)
### Packer version

1.7.5

### Simplified Packer Buildfile

<details>
<summary>sshkey.pkr.hcl</summary>

```hcl
packer {
  required_plugins {
    sshkey = {
      version = ">= 0.1.0"
      source = "github.com/ivoronin/sshkey"
    }
  }
}

data "sshkey" "packer" {}

source "null" "test" {
#  cd_content = {
#    meta-data = "instance-id: \"${uuidv4()}\""
#    user-data = <<-EOF
#      #cloud-config
#      ssh_authorized_keys:
#        - ${data.sshkey.packer.public_key}
#      EOF
#  }
#  cd_label             = "cidata"
  communicator         = "ssh"
  ssh_host             = "localhost"
  ssh_private_key_file = data.sshkey.packer.private_key_path
  ssh_username         = "root"
}

build {
  sources = ["source.null.test"]
}
```

</details>

### Operating system and Environment details

Windows 10 x64 20H2, but it doesn't  really matter

### Log Fragments and crash.log files

Include appropriate log fragments. If the log is longer than a few dozen lines,
please include the URL to the [gist](https://gist.github.com/) of the log or
use the [Github detailed format](https://gist.github.com/ericclemmons/b146fe5da72ca1f706b2ef72a20ac39d) instead of posting it directly in the issue.

Set the env var `PACKER_LOG=1` for maximum log detail.

<details>
<summary>PACKER_LOG=1 packer validate sshkey.pkr.hcl</summary>

```
2021/09/28 19:38:47 [INFO] Packer version: 1.7.5 [go1.17.1 windows amd64]
2021/09/28 19:38:47 [TRACE] discovering plugins in C:\Program Files\bin
2021/09/28 19:38:47 [TRACE] discovering plugins in C:\Users\re41236\AppData\Roaming\packer.d\plugins
2021/09/28 19:38:47 [TRACE] discovering plugins in .
2021/09/28 19:38:47 [INFO] PACKER_CONFIG env var not set; checking the default config file path
2021/09/28 19:38:47 [INFO] PACKER_CONFIG env var set; attempting to open config file: C:\Users\re41236\AppData\Roaming\packer.config
2021/09/28 19:38:47 [WARN] Config file doesn't exist: C:\Users\re41236\AppData\Roaming\packer.config
2021/09/28 19:38:47 [INFO] Setting cache directory: C:\projects\FOCUS\automation-templates\packer_cache
e: cannot determine if process is in background: Process background check error: not implemented yet
2021/09/28 19:38:47 [TRACE] listing potential installations for "github.com/ivoronin/sshkey" that match ">= 0.1.0". plugingetter.ListInstallationsOptions{FromFolders:[]string{"C:\\Program Files\\bin\\packer.exe", ".", "C:\\Users\\re41236\\AppData\\Roaming\\packer.d\\plugins"}, BinaryInstallationOptions:plugingetter.BinaryInstallationOptions{APIVersionMajor:"5", APIVersionMinor:"0", OS:"windows", ARCH:"amd64", Ext:".exe", Checksummers:[]plugingetter.Checksummer{plugingetter.Checksummer{Type:"sha256", Hash:(*sha256.digest)(0xc00081e900)}}}}
2021/09/28 19:38:47 [TRACE] Found the following "github.com/ivoronin/sshkey" installations: [{C:\Users\re41236\AppData\Roaming\packer.d\plugins\github.com\ivoronin\sshkey\packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe v0.1.0}]
2021/09/28 19:38:47 found external [-packer-default-plugin-name-] datasource from sshkey plugin
2021/09/28 19:38:47 [TRACE] Starting external plugin C:\Users\re41236\AppData\Roaming\packer.d\plugins\github.com\ivoronin\sshkey\packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe start datasource -packer-default-plugin-name-
2021/09/28 19:38:47 Starting plugin: C:\Users\re41236\AppData\Roaming\packer.d\plugins\github.com\ivoronin\sshkey\packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe []string{"C:\\Users\\re41236\\AppData\\Roaming\\packer.d\\plugins\\github.com\\ivoronin\\sshkey\\packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe", "start", "datasource", "-packer-default-plugin-name-"}
2021/09/28 19:38:47 Waiting for RPC address for: C:\Users\re41236\AppData\Roaming\packer.d\plugins\github.com\ivoronin\sshkey\packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe
2021/09/28 19:38:47 packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe plugin: 2021/09/28 19:38:47 Plugin port range: [10000,25000]
2021/09/28 19:38:47 packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe plugin: 2021/09/28 19:38:47 Plugin address: tcp 127.0.0.1:10000
2021/09/28 19:38:47 packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe plugin: 2021/09/28 19:38:47 Waiting for connection...
2021/09/28 19:38:47 Received tcp RPC address for C:\Users\re41236\AppData\Roaming\packer.d\plugins\github.com\ivoronin\sshkey\packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe: addr is 127.0.0.1:10000
2021/09/28 19:38:47 packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe plugin: 2021/09/28 19:38:47 Serving a plugin connection...
2021/09/28 19:38:47 packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe plugin: 2021/09/28 19:38:47 [TRACE] starting datasource -packer-default-plugin-name-
2021/09/28 19:38:47 [TRACE] Starting internal plugin packer-builder-null
2021/09/28 19:38:47 Starting plugin: C:\Program Files\bin\packer.exe []string{"C:\\Program Files\\bin\\packer.exe", "plugin", "packer-builder-null"}
2021/09/28 19:38:47 Waiting for RPC address for: C:\Program Files\bin\packer.exe
2021/09/28 19:38:47 packer.exe plugin: [INFO] Packer version: 1.7.5 [go1.17.1 windows amd64]
2021/09/28 19:38:47 packer.exe plugin: [INFO] PACKER_CONFIG env var not set; checking the default config file path
2021/09/28 19:38:47 packer.exe plugin: [INFO] PACKER_CONFIG env var set; attempting to open config file: C:\Users\re41236\AppData\Roaming\packer.config
2021/09/28 19:38:47 packer.exe plugin: [WARN] Config file doesn't exist: C:\Users\re41236\AppData\Roaming\packer.config
2021/09/28 19:38:47 packer.exe plugin: [INFO] Setting cache directory: C:\projects\FOCUS\automation-templates\packer_cache
2021/09/28 19:38:47 packer.exe plugin: args: []string{"packer-builder-null"}
2021/09/28 19:38:47 packer.exe plugin: Plugin port range: [10000,25000]
2021/09/28 19:38:47 packer.exe plugin: Plugin address: tcp 127.0.0.1:10000
Error: 1 error(s) occurred:
2021/09/28 19:38:47 Received tcp RPC address for C:\Program Files\bin\packer.exe: addr is 127.0.0.1:10000
2021/09/28 19:38:47 packer.exe plugin: Waiting for connection...

2021/09/28 19:38:47 packer.exe plugin: Serving a plugin connection...
* ssh_private_key_file is invalid: CreateFile <unknown>: The filename, directory name, or volume label syntax is incorrect.


  on sshkey.pkr.hcl line 12:
  (source code not available)


* ssh_private_key_file is invalid: CreateFile <unknown>: The filename, directory name, or volume label syntax is incorrect.

  on sshkey.pkr.hcl line 12:
  (source code not available)

2021/09/28 19:38:47 [INFO] (telemetry) Finalizing.
2021/09/28 19:38:48 waiting for all plugin processes to complete...
2021/09/28 19:38:48 C:\Program Files\bin\packer.exe: plugin process exited
2021/09/28 19:38:48 C:\Users\re41236\AppData\Roaming\packer.d\plugins\github.com\ivoronin\sshkey\packer-plugin-sshkey_v0.1.0_x5.0_windows_amd64.exe: plugin process exited
```
</details>

**Comment by puetzk** [comment]
It seems that packer just doesn't execute datasources in validation mode:

https://github.com/hashicorp/packer/blob/29d751a17c9677e91b18e98412f379f780023370/command/validate.go#L59-L61

Therefore `data.sshkey.packer.private_key_path` is undefined during `validate` run.

However, it looks like when it skips executing, in places specific `cty.UnknownVal` placeholders.

https://github.com/hashicorp/packer/blob/a534c743cf21e9cd6bcb97367254f99767a24582/hcl2template/types.packer_config.go#L299-L301

I wonder if one could build on that to realize that these are values that (presumably) will be supplied by the datasource in a real run (maybe using https://github.com/zclconf/go-cty/blob/main/docs/marks.md), and then ignore the values of fields which are marked incomplete in that way during validation, knowing that if they are incomplete it's presumably becuase the datasource would fill in somethingin a real run. Obviously that could still fail at runtime - we don't know what the datasource will return. But it would allow validating things that are static, which is better than nothing.

**Comment by lmayorga1980** [comment]
I am experiencing the same issue trying to use dynamic ssh keys with the same plugin.

**Comment by puppetmaster** [comment]
> 

+1 here

**Comment by jbrongust** [comment]
I think i'm running into the same issue but when trying to use a local variable built from `amazon-secretsmanager` data value.

I think this issue describes it well: https://github.com/hashicorp/packer/issues/11197

```
data "amazon-secretsmanager" "super-secret" {
  name = "mysecrets/dont-tell"
}

locals { 
  timestamp = regex_replace(timestamp(), "[- TZ:]", "") 
  client_secret = data.amazon-secretsmanager.super-secret.value
  }

build {
  sources = ["source.amazon-ebs.autogenerated_1"]
    provisioner "shell" {
      environment_vars = ["CLIENT_SECRET= ${local.client_secret}"]
      remote_folder    = "${var.remote_folder}"
      script           = "${path.root}/scripts/master_plan.sh"
    }
}
```

Used the work around described here: https://github.com/hashicorp/packer/issues/11795#issuecomment-1147482449

**Comment by lbajolet-hashicorp** [comment]
Hi there,

First off, sorry for letting that one slide, it's been opened for a while and went under our radar, apologies.

This problem is slightly harder to solve than handling `cty.Unknown` types unfortunately, as plugins are in charge of validating the values they receive, and that's what `packer validate` relies on to determine whether or not a template is valid.

As such, these plugins will do what they would for a normal run: check that their configuration is OK statically. As you pointed out @puetzk, this may still fail at runtime for a lot of reasons, but at least conceptually, the ground is all set for the plugin to work.

Because of this model, we send a bunch of information over the wire to the plugins, and they don't see `cty` values, they see plain `go` values. So in your example, the path received by the plugin's `Validate` function will be `"unknown"`, hence the `ssh_private_key_file` error you get, not being a valid file.

Now the way we see it, there are two possible workarounds:

1. Define all the problematic variables with local values, and set them as placeholders for validation, then change it to the datasource-powered ones. This implies that the template you're validating is replaced by the appropriate values at run time, but this can let you at least verify that the structure of the template is OK.
2. This is what is solved by #12106: we evaluate the datasources, so the plugins have real-data to validate the configuration with.

The solution 2 in your case will likely be optimal, as you only rely on local sources, so this will be fast, and cost-free. For those plugins with a datasource that relies on external services, this may incur a cost, hence why we don't opt for this to be the default behaviour, but this is given as an option if you want to validate the template before running it.

Hope these solutions will solve your problem, and that the approach is satisfactory for everyone!

Thanks for reporting, and sorry again for how long we've taken to answer this!
